tooltips: Add a dedicated tooltip for status-emojis.

shashank-23002 opened the pull request:
This issue requires us to add a dedicated tooltip whenever the user hovers over the status-emoji in the left and right sidebar.Currently we don't show the name of status-emoji in sidebars and only show the active-status tooltip.

This PR adds a dedicated tooltip in left and right sidebar whenever user hovers over the status-emoji.
Provisionally, the changes done are - 
1) Whenever the status-emoji is hovered in right sidebar , the buddy list tooltips instance will be hidden and vice-versa for status-emoji-tooltip.
2) Same goes for left sidebar / dm section . 

Fixes: #26772

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<img width="606" alt="Screenshot 2024-01-19 at 8 36 20 PM" src="https://github.com/zulip/zulip/assets/97145463/b7256944-b3fa-45e7-a3ca-37da81ed9a3a">
<img width="606" alt="Screenshot 2024-01-19 at 8 36 41 PM" src="https://github.com/zulip/zulip/assets/97145463/982209aa-c7b9-4c62-b773-bfeb2c807f77">


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
You will need to add screenshots, go through the self-review checklist, and fix failing tests in order for this PR to be ready for review.

shashank-23002 commented:
Hey @alya I have added screenshots and fixed the failing tests. PTAL

alya commented:
Ah, but my comment had three points to take care of. ;)

shashank-23002 commented:
Thanks @alya I have updated the self-review checklist for this PR to be ready for review.

alya commented:
@karlstolley do you want to do a first-round review on this one? I haven't tested it.

karlstolley commented on web/templates/status_emoji.hbs:
```diff
@@ -6,7 +6,7 @@
-{{else if emoji_code}}
-<span class="emoji status-emoji emoji-{{emoji_code}}"></span>
+{{else if emoji_name}}
+<span class="emoji status-emoji emoji-{{emoji_code}}" id="status-emoji-name" data-tippy-content=":{{emoji_name}}:"></span>
```
Let's handle `status-emoji-name` as a class; an `id` should only appear once in the DOM.

karlstolley commented on web/src/click_handlers.js:
```diff
@@ -565,6 +567,10 @@ export function initialize() {
         }
 
+        if (user_status.get_status_text(user_id)) {
+            return;
```
These early returns are non-starters--if a user has a status set, the regular buddy-list tooltip will not be rendered. And that's a significant regression.

karlstolley commented:
Thanks for your work here, @shashank-23002 .See my comments above--the approach here introduces a significant regression, because the ordinary user tooltip will not be displayed if a user has a status set.

What needs to happen is a better separation of which username elements get the user tooltip, and how to isolate that from the status emoji, which should have its tooltip set on a specific selector.

shashank-23002 commented:
Thanks @karlstolley I have fixed the regression and updated the id with class.
But the main concern due to which it has become a little confusing to resolve this issue is that status-emoji selector lies inside the username tooltip selector due to which both of them are triggered simultaneously.
Moreover when I change the target element to isolate the status-emoji selector from the username tooltip selector , the username tooltip selector becomes limited to just the user-name class and the tooltip won't be triggered for the status-text as it was happening before.

shashank-23002 commented:
Hey @karlstolley I have done the necessary changes which can fix this issue. PTAL

karlstolley commented:
Thanks for your patience and your work here, @shashank-23002! The code changes here are looking good, and manual testing of the feature in the left and right sidebars and compose pills all acted as expected. I'll mark for QA review so that @alya has a chance to poke at it, but I suspect this will soon be ready for integration review.

shashank-23002 commented:
Thanks Karl , would also like to mention that probably this could close #19821 because the current issue actually originated from the former. 
Would be really great if both of the issues are closed simultaneously.

alya commented:
Thanks! Unfortunately, this isn't quite working right.

If I hover over the status emoji, and then move my mouse to the user's name, the regular tooltip for the user should get displayed; it currently does not. Going in the other direction (name -> emoji) seems to work as expected.

alya commented:
Also, with some amount of playing with it, I was able to get both tooltips to appear at the same time, which I think should never happen. I'm not sure how to reproduce this.

![Screenshot 2024-01-04 at 2 20 02 PM](https://github.com/zulip/zulip/assets/2090066/9533a7d8-6726-4a6d-a67f-e3f2bc21738c)

alya commented:
Also, it looks like this feature does not work for custom emoji. Is that expected?

shashank-23002 commented:
Thanks @alya for reviewing the code.

The logic I am using here is to hide all the tooltips when hovering over the status-emoji and to just show the emoji-tooltip. 
Now , to trigger the user-name tooltip once again we need to re-enter the trigger defined in user-name tooltip i.e ```mouseenter on selectable-sidebar-block```.

As we can see the status-emoji is a part of user-name tooltip trigger so they can't be separated apart. Thus it is obvious that once the user-name tooltip has been hidden , you need to re-trigger it through the defined logic.
One logic can be to store all the tooltips which were hidden when status-emoji-tooltip was triggered and reshow all of them when the user leaves. But that makes it a little heavy when the user just keep entering and re-entering the status-emoji.

So, I guess probably the current logic might not cause an issue for the user to re trigger the user-name tooltip when he has already seen one before hiding from the status-emoji tooltip.

shashank-23002 commented:
> Also, with some amount of playing with it, I was able to get both tooltips to appear at the same time, which I think should never happen. I'm not sure how to reproduce this.
> 
> ![Screenshot 2024-01-04 at 2 20 02 PM](https://private-user-images.githubusercontent.com/2090066/294312876-9533a7d8-6726-4a6d-a67f-e3f2bc21738c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ0NTA2NjUsIm5iZiI6MTcwNDQ1MDM2NSwicGF0aCI6Ii8yMDkwMDY2LzI5NDMxMjg3Ni05NTMzYTdkOC02NzI2LTRhNmQtYTY3Zi1lM2YyYmMyMTczOGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMDVUMTAyNjA1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2ZmYTMzMjNjMjE0NjJiM2Q4NTg4MGYzYWJjYzU0MWY4MjJlNmRmODMwN2U4MmQyMzQwM2I0ZjJkYzQ3YTdlZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.B-DuJJfxWcJGtag_jHKtdvP5HSTM43WczdI7YUnexRs)

I tried to reproduce the issue but couldn't figure out if it would be possible. Because according to the logic , once the status-emoji-tooltips are shown all the other tooltip instances are hidden.
Might be an issue with slow-network due to which the tooltips didn't hide instantly.

Would love to hear from @karlstolley if he faced the same issue while reviewing or for any suggestions.
Thanks.

shashank-23002 commented:
> Also, it looks like this feature does not work for custom emoji. Is that expected?

Ohh thanks , I didn't have the idea about this feature. 
Will update my PR accordingly

alya commented:
> So, I guess probably the current logic might not cause an issue for the user to re trigger the user-name tooltip which he already does everytime he needs to see the user-list tooltip. Moreover which he has already seen before hiding from the status-emoji tooltip.

I still think it would be better for the tooltip to appear, but I'll let @karlstolley comment on whether this is too tricky.

karlstolley commented on web/src/tippyjs.ts:
```diff
@@ -502,10 +503,20 @@ export function initialize(): void {
+        delay: INSTANT_HOVER_DELAY,
+        appendTo: () => document.body,
+        onShow() {
+            hide_all();
```
So I think this is not the right approach; [Tippy has its own hideAll() method](https://atomiks.github.io/tippyjs/v6/tippy-instance/) we could use. I think the better approach here would be to find [the Tippy instance](https://atomiks.github.io/tippyjs/v6/tippy-instance/) shown on the username, and then [hide](https://atomiks.github.io/tippyjs/v6/methods/#hide) it while the status emoji tooltip is being shown. Once the status emoji tooltip is hidden, then again [show](https://atomiks.github.io/tippyjs/v6/methods/#show) the username Tippy.

shashank-23002 replied:
Thanks @karlstolley , I thought about the above approach as well but the main issue with the above was to locate the instance associated with the user-name in the .status-emoji-name delegate function so that I can hide it. This is because the tooltip is shown for the complete ```user-presence-link``` which includes user-name as well as status-emoji.
Was a little busy with college stuffs so couldn't look up. Will try to fix this ASAP.

shashank-23002 commented:
@karlstolley I have changed my approach and updated the required changes as per your suggestions. 
Now the tooltips appear as @alya expected to have them.
Hope that works.

alya commented:
Works for me in manual testing, including custom emoji!

alya commented:
@karlstolley please mark this PR for integration review whenever it's ready from your perspective.

karlstolley commented:
Thanks @shashank-23002! This looks like a solid set of changes.

I think the next step is to give this some burn-in time on CZO, but I'll just mark for integration review so @timabbott can decide what's best.

karlstolley commented:
Though @shashank-23002 please use the correct image markdown to display your screenshots, rather than make reviewers follow links.

shashank-23002 commented:
Thanks for reviewing the changes Karl . 

I have updated the screenshots , though I couldn't understand why am I unable to upload screen recordings on this PR. Whenever I do so , it converts into followable links.
Hope there's not an issue with current screenshots.

timabbott commented on web/src/tippyjs.ts:
```diff
@@ -5,6 +5,7 @@ import tippy, {delegate} from "tippy.js";
 
 import {$t} from "./i18n";
+import {hide_all} from "./popovers";
 import {user_settings} from "./user_settings";
```
I know the two other imports in this file work differently, but please import this as `import * as popovers from "./popovers";` and access it as `popovers.hide_all`. Those two other modules are special; 98% of modules should be imported the `*` way so the namespacing is clear.

shashank-23002 replied:
Updated . Hope that works .

timabbott commented on web/src/click_handlers.js:
```diff
@@ -573,6 +573,24 @@ export function initialize() {
             check_reference_removed,
         );
+
+        $(".status-emoji-name").off("mouseenter").off("mouseleave");
```
I notice that the commit message is missing a blank line between the title and body. Every git commit should have such a blank line.

shashank-23002 replied:
Should be good to go now.

timabbott commented:
I've test-deployed this on chat.zulip.org; we should bump the relevant thread for testing/feedback, since I doubt folks will test this a lot by accident.

karlstolley commented:
[CZO discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/viewing.20emoji.20names.20of.20status.20emoji) -- @shashank-23002, it would be good to link to that from the

shashank-23002 commented:
> [CZO discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/viewing.20emoji.20names.20of.20status.20emoji) -- @shashank-23002, it would be good to link to that from the original pull request text.
> 
> Sorry about the closing and reopening--my palm landed on my trackpad.

Sure , I bumped a thread on [CZO](https://chat.zulip.org/#narrow/stream/137-feedback/topic/view.20status-emoji-names) for feedback and updated my PR fixing an issue faced by a user.

timabbott commented on web/src/tippyjs.ts:
```diff
@@ -502,10 +503,24 @@ export function initialize(): void {
-        target: "#user_info_popover .status-emoji",
+        target: ".user-card-status-text .status-emoji",
         appendTo: () => document.body,
     });
```
What's this change do?

timabbott replied:
OK -- can you put it in its own commit (either a prep commit or new final commit)? I think a commit just after that one renaming that block from `user-card-status-text` to `user-card-status-area` would be a big help; the current name really hints (incorrectly) that it's just for a textual status, and not emoji.

Check out our [GitHub guide](https://zulip.readthedocs.io/en/latest/git-guide.html) and [commit guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for details on why this is important and really helpful.

shashank-23002 replied:
Sure thing. Even I found it a little vague to add the `user-card-status-text` class but I was unsure.
Updated the changes in different commits.

timabbott commented on web/src/tippyjs.ts:
```diff
@@ -502,10 +503,24 @@ export function initialize(): void {
+        onHidden(instance) {
+            instance.destroy();
+        },
+    });
```
So this solution covers all the locations other than the sidebars; can you add some comments both here and in `click_handlers.js` mentioning that the other code paths exist, and explaining why we need the sidebar-specific logic?

I think what you have is probably correct but we should document it for posterity.

shashank-23002 replied:
Sure , added the comments explaining why we need a different logic for sidebars.
PTAL.

timabbott commented on web/src/click_handlers.js:
```diff
@@ -573,6 +573,28 @@ export function initialize() {
+
+        /*
+           We need a separate way to handle mouseevents for status-emoji in sidebars apart from tippyjs.ts logic. This is because status-emoji is a part of buddy-list tooltip trigger and we need to separate both the tooltips to appear at the same time. Hence when we enter the emoji section, the exact buddy-list tooltip instance needs to be destroyed. Also  when we leave the emoji section , the bu ...
+        */
```
Please line-wrap your comments (as we do everywhere else in the code base) and make them more concise.

shashank-23002 replied:
Hope the present changes works.

timabbott commented:
I did some tweaks to the comments, moved the tooltip placement to "top", and marked this to merge once CI passes, thanks @shashank-23002!
